proxies.py

```diff
@@ -0,0 +1,183 @@
+        if status != 101:
+            raise Exception("Bad Response, %s, %s" % (status, headers))
+        self.do_proxy(tsock)
+
+    def do_proxy(self, target):
+        """
+        Handle the reading and writing of data being passed through
+        the proxy
+
+        Args:
+          target: the websockify WebSocketServer socket of the target
+            server
+        """
+        cqueue = []
+        tqueue = []
+        rlist = [self.request, target]
+        self.request.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        target.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+        if self.server.heartbeat:
+            now = time.time()
+            self.heartbeat = now + self.server.heartbeat
+        else:
+            self.heartbeat = None
+
+        while True:
+            wlist = []
+
+            if self.heartbeat is not None:
+                now = time.time()
+                if now > self.heartbeat:
+                    self.heartbeat = now + self.server.heartbeat
+                    # self.send_ping()
+
+            if tqueue: wlist.append(target)
+            if cqueue: wlist.append(self.request)
+            try:
+                ins, outs, excepts = select.select(rlist, wlist, [], 1)
+            except (select.error, OSError):
+                exc = sys.exc_info()[1]
+                if hasattr(exc, 'errno'):
+                    err = exc.errno
+                else:
+                    err = exc[0]
+
+                if err != errno.EINTR:
+                    raise
+                else:
+                    continue
+
+            if excepts: raise Exception("Socket exception")
+
+            if self.request in outs:
+                # Send queued target data to the client
+                dat = ''.join(cqueue)
+                cqueue = []
+                sent = self.request.send(dat)
+                if sent != len(dat):
+                    # requeue the remaining data
+                    cqueue.insert(0, dat[sent:])
+
+            if self.request in ins:
+                # Receive client data, decode it, and queue for target
+                buf = self.request.recv(self.buffer_size)
+
+                if len(buf) == 0:
+                    if self.verbose:
+                        self.log_message("%s:%s: Client closed connection",
+                                         self.server.target_host,
+                                         self.server.target_port)
+                    raise self.CClose(1000, "Target closed")
+
+                else:
+                    #Don't forward pong packets
+                    frame = self.decode_hybi(
+                        buf, base64=self.base64, logger=self.logger,
+                        strict=self.strict_mode)
+                    #ignore pongs with no payload
+                    if not (frame['opcode'] == 0xA and not frame['payload']):
+                        tqueue.append(buf)
+
+
+            if target in outs:
```

Comment:
This code is identical to the code above on lines 137-144. Could this be refactored into a function which takes the queue of data and the socket to send it to as arguments?